Bug 1465703 - Add bookmark and follow_bookmark probes for Savant Shie…

biancadanforth opened the pull request:
…ld study; r=mak, adw

Fixes #27 

TODO
- [x] Create issue detailing implementation and any limitations or additional details
- [x] Update TESTPLAN.md (Issue #5 )
- [x] Import commits into Hg and run this PR on the Try server with study pref ON and OFF (`./mach try -b o -p win64,linux64,macosx64 -u mochitests,xpcshell -t none`). Post links in this PR. Resolve issues as required.
- [x] Push to Review Board, push to Try again via RB GUI
- [x] obtain r+ from Peer (mak OR adw) and QA (pdehaan)
- [x] Land patch in Nightly

These probes will register and record (for the duration of the study only):
* When a bookmark is added.
* When a bookmark is removed.
* When a bookmark is visited.

Note for bookmark added: Using the 'onItemAdded' nsINavBookmarkObserver will not distinguish bulk adds (as from creation of default bookmarks for a new user, bookmark migration from another browser, Sync, ...) from one-off adds by the user through the UI. This observer will be used for simplicity, since there are many locations in the UI where a user can add a bookmark. More fine grained probes can be implemented later if desired.

Note for bookmark visited: This will also fire for the case when the user visits a page directly that happens to be bookmarked. The user doesn't have to "click" on a bookmark.

biancadanforth commented:
Try run: `./mach try -b o -p win64,linux64,macosx64 -u mochitests,xpcshell -t none`
* with study pref OFF: https://treeherder.mozilla.org/#/jobs?repo=try&revision=31f4f02743a622bd7a7b694218b04a01b13fd96b
* with study pref ON: https://treeherder.mozilla.org/#/jobs?repo=try&revision=dd0f9c58670561533bfc30114140e044f2029afc

mak77 commented on browser/modules/SavantShieldStudy.jsm:
```diff
@@ -176,4 +185,65 @@ class TelemetryEvents {
+  init() {
+    this.observer = {
+      onItemAdded: (itemID, parentID, index, itemType, uri, title, dateAdded, guid, parentGUID) => {
+        if (itemType === this.TYPE_BOOKMARK) {
```
You probably want to also check && !uri.schemeIs("place") to exclude special places queries.
Also in onItemRemoved

mak77 replied:
Also, note that most of these notifications have a source argument you can use to exclude bookmarks modified by: sync, import, restore.
Pretty much you could just check source == PlacesUtils.bookmarks.SOURCE_DEFAULT

biancadanforth replied:
Thanks for the comments; for the first one, what is a bookmark with a `"place"` scheme for? When would a `place://...` URI bookmark be added? Just trying to understand what the check is about.

That is a great suggestion for the second point; I have added that check. Looks like the docs in MDN are not up to date? I don't see a "source" argument there for [onItemAdded](https://developer.mozilla.org/en-US/docs/Mozilla/Tech/XPCOM/Reference/Interface/nsINavBookmarkObserver#onItemAdded()) or [onItemRemoved](https://developer.mozilla.org/en-US/docs/Mozilla/Tech/XPCOM/Reference/Interface/nsINavBookmarkObserver#onItemRemoved()).

mak77 replied:
MDN is terribly out-of-date, the .idl inline documentation is usually much better :(

"place:query" uris (actually without //) are special uris for internal use that allow to create containers like Most Visited or Recently Bookmarked. We add a couple of those as default bookmarks, they should be pretty rare and it's unlikely for users to create them manually.
Filtering them out doesn't hurt, but even not doing that is unlikely to skew your stats since they are rare.

mak77 commented on browser/modules/SavantShieldStudy.jsm:
```diff
@@ -176,4 +185,65 @@ class TelemetryEvents {
+  }
+
+  init() {
+    this.observer = {
```
You may want to add skipTags: true, to the observer object, so you won't be notified of "fake" bookmarks we currently create for tags.

biancadanforth replied:
Yes I do want that! Thank you; added.

mak77 reviewed (commented):
Looks good overall, there's a bit more filtering you may apply to notifications.
Note this will have a small effect on performance, because you are effectively adding a new bookmark changes listener, but as far as this experiment is time limited it will be ok.

biancadanforth commented:
@mak77 , One thing I observed in [Treeherder](https://treeherder.mozilla.org/#/jobs?repo=try&revision=dd0f9c58670561533bfc30114140e044f2029afc&selectedJob=181850490), when the study pref is ON, are consistent test failures for the "browser/base/content/test/performance/browser_startup.js" test, citing that "resource://gre/modules/Bookmarks.jsm is not allowed before handling user events -".

Should I wait to initialize my `BookmarkObserver` class until a notification like `sessionstore-windows-restored` or `browser-delayed-startup-finished` have fired?

mak77 commented:
It may be a good idea to start your module **after** sessionstore-windows-restored. You may call your Init() in _scheduleStartupIdleTasks, using Services.tm.idleDispatchToMainThread.

biancadanforth committed:
Bug 1465703 - Add bookmark and follow_bookmark probes for Savant Shield study; r=adw, a=RyanVM

NOTE: THIS PATCH DOES NOT WORK FOR BOOKMARK ADD/REMOVE UNTIL A FIX IS APPLIED IN THE TAB PROBE PATCH.

These probes will register and record (for the duration of the study only):
* When a bookmark is added.
* When a bookmark is removed.
* When a bookmark is visited.

Also moved init of study module to later in startup as an idle task to mitigate performance impact. Previously, this patch was failing the browser/base/content/test/performance/browser_startup.js test.

Note for bookmark added/removed: Using the source argument, we can filter out bulk events like from Sync, import, restore, etc.

Note for bookmark visited: This will also fire for the case when the user visits a page directly that happens to be bookmarked. The user doesn't have to "click" on a bookmark.

MozReview-Commit-ID: EYVlTLfVLkd
